KEP-2578: Windows Operational Readiness Specification (#2975)
kubernetes/enhancements: keps/sig-windows/2578-windows-conformance/README.md

This KEP first abstractly defines the various categories of Windows "conformance", differentiating between core conformance from feature sets that are of common interest.

Next, we define how these conformance tests can be run using *existing* Kubernetes ginkgo tags, without any additional tooling.

Finally, we prescribe the implementation of a generic "windows conformance" testing tool, which can be either hand-rolled (i.e. in a bash script) or, implemented as part of a broader initiative (i.e. inside of the sonobuoy.io, which is commonly used for Kubernetes verification).

The initial implementation of this KEP is mainly focused on standardizing Windows Core Conformance definitions.  Other features, such as `HostProcess`, will be refined and updated with new releases, as these APIs and their standardization evolve in the community.

## Motivation

> Running Windows containers at scale is a requirement for many mature enterprises migrating applications Kubernetes.  The requirements of a Windows Kubelet are not comparable to a Linux Kubelet, however, because of the fact that Windows does not support the same storage, security, and privileged container model that Linux does.  We thus need an unambiguous standard in place that is meaningfull and actionable for comparing Windows Kubernetes support between vendors.

Reviewer: comments from @jdumars @perithompson 
- need OS specifics
- need backward compatibilty stuffsss

Reviewer 2: ```suggestion
Running Windows containers at scale is a requirement for many mature enterprises migrating applications Kubernetes.  The requirements of a Windows Kubelet are not comparable to a Linux Kubelet, however, because of the fact that Windows does not support the same storage, security, and privileged container model that Linux does.  We thus need an unambiguous standard in place that is meaningful and actionable for comparing Windows Kubernetes support between vendors.
```

- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

We aim to differentiate Operation Conformance from Kubernetes Certified Conformance in this KEP as a mechanism to solve the end user problem of end users needing to run Windows workloads and user Kubernetes in a predictable way, defined in two dimensions:

> - Clusters capable of running Windows containers also must be capable of passing Linux Conformance tests, by leveraging available Linux nodes - for example - by levaraging the Linux control plane which is omnipresent in any K8s cluster.

Reviewer 2: ```suggestion
- Clusters capable of running Windows containers also must be capable of passing Linux Conformance tests, by leveraging available Linux nodes - for example - by leveraging the Linux control plane which is omnipresent in any K8s cluster.
```

> - Clusters capable of running Windows, with versinos > 2019 and forward, conform to the Operations tests defined and standardized in this document.

Reviewer 2: ```suggestion
- Clusters capable of running Windows, with versions > 2019 and forward, conform to the Operations tests defined and standardized in this document.
```

> In general, the ad-hoc usage of bespoke ginkgo tags is cubmbersome and makes adoption of Windows workloads on Kubernetes difficult to certify.  In addition to defining these specific definitional constructs for Windows conformance in this KEP, we also aim to implement these defintitions in a declarative, fully automated certification bundle via the sonobuoy (https://sonobuoy.io/) tool, as a quick and easy way for end-users to verify windows clusters.

Reviewer 2: ```suggestion
In general, the ad-hoc usage of bespoke ginkgo tags is cumbersome and makes adoption of Windows workloads on Kubernetes difficult to certify.  In addition to defining these specific definitional constructs for Windows conformance in this KEP, we also aim to implement these definitions in a declarative, fully automated certification bundle via the sonobuoy (https://sonobuoy.io/) tool, as a quick and easy way for end-users to verify windows clusters.
```

### Goals

- Disambiguate the verification requirements for Windows Conformance testing in Kubernetes from the implementation of tests themselves.
- Specify a canonical mechanism for validating Windows Conformance which leverages the existing e2e test suite in a declarative way.
- Standardize the definitions for how we verify support of new Windows features in Kubernetes, and standardize the features *not* fully supported by Windows.

### Non Goals

- Modify the existing Kubernetes Conformance specification, which has linux specific tests.
- Modify existing tests inside of Kubernetes that would increase the bifurcation of testing implementation.
- Adding new tooling to Kubernetes core itself for testing Windows.

## Proposal

### Background

Currently, tools such as the Kubernetes end-to-end testing suite and Sonobuoy can be used to verify Kubernetes clusters, but the filters and, more importantly, the meaning behind certain tests can be ambiguous.

> If we consider the various idiosyncracies of verifying Kubernetes on Windows we are immediately confronted with several potential hotspots that can be difficult to wade through as an end user.

Reviewer 2: ```suggestion
If we consider the various idiosyncrasies of verifying Kubernetes on Windows we are immediately confronted with several potential hot-spots that can be difficult to wade through as an end user.
```

- Some tests (including Networking, NetworkPolicy, and Storage related tests) behave differently on Windows by necessity, because of differences in the feature sets and testing idioms that are supportable on Windows Server.  We aim to distinguish the core requirements of such tests in terms of conformant Windows clusters, irrespective of implementation gaps which may exist between how Linux and Windows implement certain tests.
- Some critical windows functionality, such as the ability to run activeDirectory as an identity provider for a pod, need to be well-defined in a Kubernetes context for individuals planning on integrating production Windows workloads into their Kubernetes support model.
- The introduction of "Privileged containers" (formally referred to as `HostProcess` pods), further differentiates Linux from Windows in subtle ways which specifically will, eventually, allow Windows users to implement many of the common idioms in Kubernetes using a comparable, but different runtime paradigm.

> - The Kubernetes ServiceProxy implementation is not at parity with that of Linux, and likely never will be.  This needs to be made explicit so that attempts to reach parity between Kuberentes service API implementations on Windows and Linux can be approached in a systematic manner in terms of core requirements and supportability.  Features such as DSR (direct-server-return), EndpointSlices, and so on generally require special treatment for Windows clusters, some of which may have a complex interplay with the underlying CNI provider.

Reviewer 2: ```suggestion
- The Kubernetes ServiceProxy implementation is not at parity with that of Linux, and likely never will be.  This needs to be made explicit so that attempts to reach parity between Kubernetes service API implementations on Windows and Linux can be approached in a systematic manner in terms of core requirements and supportability.  Features such as DSR (direct-server-return), EndpointSlices, and so on generally require special treatment for Windows clusters, some of which may have a complex interplay with the underlying CNI provider.
```

We propose the following definitions as the Windows conformance standard for Kubernetes clusters that are released after Kuebrentes 1.23 and onwards.  This specification is to be updated as new features are added, or removed, from the Kubernetes API itself.

### User Stories (Optional)

#### Story 1

> As a Windows Application developer I want to verify that the features i rely on in Kubernetes are supported on my specific K8s Windows cluster.

Reviewer 2: ```suggestion
As a Windows Application developer, I want to verify that the features I rely on in Kubernetes are supported on my specific K8s Windows cluster.
```

#### Story 2

> As a Kubenetes vendor I want to evalaute the completeness of my Kubernetes support matrix in context of Windows supported features in the Kubernetes API.

Reviewer 2: ```suggestion
As a Kubernetes vendor I want to evaluate the completeness of my Kubernetes support matrix in context of Windows supported features in the Kubernetes API.
```

#### Story 3

> As a Kubernetes developer I want to immediately know wether or not the features I want to add will effect the Windows support matrix, and if so, I want to be able to rapidly convey this in a canonical way to the broader community.

Reviewer 2: ```suggestion
As a Kubernetes developer I want to immediately know whether or not the features I want to add will affect the Windows support matrix, and if so, I want to be able to rapidly convey this in a canonical way to the broader community.
```

#### Story 4

> As an IT administrator, I want to know wether my current version of Kubernetes supports ActiveDirectory

Reviewer 2: ```suggestion
As an IT administrator, I want to know whether my current version of Kubernetes supports ActiveDirectory
```

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Specification of Windows Core Conformance (Kubernetes 1.23)

- In order to qualify as a Windows test, the pods that demonstrate functionality for a given attribute of a cluster much be running on a `pause` image, and this image must be scheduled to an operating system that:

    - 1809

>     - 1909

Reviewer 2: Why are we considering 1909 and 2004 for Windows Conformance? 1909 SAC has already reached it's end of service and 2004 SAC is rapidly approaching it's end of service at the end of this year. From what I understand, Microsoft is dropping SAC versions entirely in the coming future and will be LTSC only.

I believe we should only be performing conformance tests on what would be officially supported at the time of 1.23, which would be LTSC versions of Windows Server (Windows Server 2019 and Windows Server 2022).

Furthermore, we should consider switching to identifying Windows Server 2019 as something other than 1809 as this will continue to be a source of confusion as this build version refers to both an LTSC and SAC release.

Reviewer: per conversation in sig-windows @perithompson / ross to push updates for this 

2 latest SACs
latest LTS

Reviewer: 2004 -> out per muzz

Reviewer: curate a table as part of this KEP , updated from version to version

Reviewer 3: Do we need to have some way to define Windows OS patches that are required for Operational Conformance?

Reviewer: removed since 2022 simplifies things

- Ability to mount a hostPath storage volume from a Kubelet into a pod and write to it. 
- Ability to mount a hostPath storage volume from a Kubelet into a pod and read from it. 
- Ability to schedule multiple containers, with distinct IP addresses, on the same node.
- Ability to reboot all of the Windows nodes in a cluster while running windows pods, in such a way that recovery of these deleted pods occurs in the scope of a few minutes.
- Ability to delete and recreate services in such a way that load balancing rules for pods are recovered by whatever chosen service proxy is being utilized.
- Ability to delete and recreate pods for StatefulSets which preserve their ability to serve as routed endpoints for services, while also having unchanging IP addresses.
- Ability to access internal pods by internal Kubernetes DNS service endpoints.
- Ability to access external services by Kubernetes DNS services endpoints (for non-airgapped clusters).
- Ability to access internal pods through custom DNS records, injected by the Kubernetes pod Specification.
- Ability to route services from pods from the EndpointSlice API.

> 

Reviewer 3: ```suggestion
- Ability to run IPv4/IPv6 dual-stack networking (on supported OS Versions)
```

#### Basic Service accounts and local storage

- Ability to rotate service accounts for running pods, with the APIServer access capacity of the pod remaining unchanged over the long term.
- Ability to read and write shared files on a single Kubernetes node between three running Windows containers, simultaneously.

#### Basic Scheduling

- Ability to schedule pods with CPU and Memory limits demonstrably honored over time.
- Ability to demonstrate the pods requesting more CPU and Memory then available are left in the Pending state over time.

#### Basic performance

> - Ability to reschedule the a deployment Pod, 50 times in a row, with continuous deletions in the background.

Reviewer 2: ```suggestion
- Ability to reschedule a deployment Pod, 50 times in a row, with continuous deletions in the background.
```

- Ability to route traffic to 10 pods behind a common ClusterIP service endpoint.
- Ability to route traffic to 10 pods behind a common NodePort service endpoint.

> We do not include GMSA support in the basic definition of WindowsConformance, because it is known to require other cluster components outside of Kubernete's Control.

Reviewer 2: ```suggestion
We do not include GMSA support in the basic definition of Windows Conformance because it is known to require other cluster components outside of Kubernetes's Control.
```

### Sub-specifications of Windows Conformance (Kubernetes 1.23)

> The following subsets of verifications for Windows clusters further expand the way we define supportability for Windows clusters, but they are not considered part of the "core" Conformance specification.  These features may be *heavily* reliant on functionality which resides well outside of Kubernete's preview, for example: CNI Implementation, Container Runtime implementation, or Windows Server Edition + Configuration.

Reviewer 2: ```suggestion
The following subsets of verification for Windows clusters further expand the way we define supportability for Windows clusters, however, they are not considered part of the "core" Conformance specification.  These features may be *heavily* reliant on functionality which resides well outside of Kubernetes's purview, for example: CNI Implementation, Container Runtime implementation, or Windows Server Edition and Configuration.
```

#### Windows HostProcess Conformance

Note that, HostProcess Containers have not yet been verified to support all of these features.

> Nevertheless, we define this specification because it is an obvious starting point for the long-term, meaningfull support of a full-fledged HostProcess feature in Kubernetes.

Reviewer 2: ```suggestion
Nevertheless, we define this specification because it is an obvious starting point for the long-term, meaningful support of a full-fledged HostProcess feature in Kubernetes.
```

- Ability to access the APIServer using pod mounted service accounts from a hostProcess pod.
- Ability to create and manage host level networking (hcn) rules from a Windows hostProcess pod.
- Ability to launch hostProcess containers which share the IP address of a Windows node.
- Ability to launch hostProcess containers which can run other privileged Windows system API (to be specified further in the future).

> - Ability to utilize NT Authority/Network Service should be able to interact with host network objects

Reviewer 3: I don't dont think this actually works today, it seems to need NT Authority/System?

Reviewer: ack, ok removed

#### Active Directory

- Ability to run a pod as a GSMA User with only be able to access allowed resources
- Ability to read and write from local storage using ActiveDirectory credentials
- Observed lack of Ability to read and write from local storage when ActiveDirectory protected resources credentials are not present

#### Network Policies

Note that we do not include UDP, ipv6, in the initial definitions for NetworkPolicy conformance on Windows.  Since NetworkPolicy testing support on

> Windows clusters was just recently added to Kuberentes Core, we err on the side of a conservative definition for Windows NetworkPolicy which

Reviewer 2: ```suggestion
Windows clusters was just recently added to Kubernetes Core, we err on the side of a conservative definition for Windows NetworkPolicy which
```

is sufficient to address the needs of *most* legacy network security applications.

- Ability to protect IPv4 pods from acceessing other pods when TCP NetworkPolicys are present that block specific pod connectivity.
- Ability to protect IPv4 pods from acceessing other pods when TCP NetworkPolicys are present that block specific namespace connectibity.

#### Windows Service Proxying

- Ability to support IPv6 and IPv4 services for accessing internal Linux pods from Windows pods over ClusterIP Endpoints.
- Ability to support IPv6 and IPv4 services for accessing internal Linux pods from Windows pods over NodePort Endpoints.
- Ability to support IPv6 and IPv4 services for accessing internal Windows pods from Linux pods over ClusterIP Endpoints.

> - Ability to support IPv6 and IPv4 services for accessing internal Windows pods from Linux pods over NodePort Endpoints.

Reviewer 3: Is this only on some of the Windows versions?

Reviewer 4: IPv6 dual stack is only supported on 2004+

Reviewer: clarified

### Implementation

This section details the implementation of a "tool" which verifies these clusters must allow for the following, differentiated set of verifications, specified above.

- core conformance
- hostProcess
- activeDirectory
- networkPolicy
- serviceProxy

#### Option 1 - Test tags

One way to implement such a tool is to, specifically, target all available "ginkgo" tests in Kubernetes core itself, and run these one at a time (or concurrently).  Not we *do not* assert that these tests *must* run in parallel in order to comprise a conformance clusters (although this is often done in CI, its generally not a requirement of Kubernetes Conformance for linux clusters, either, since cluster capacity is not related to the functionality of a cluster). As an example, one such implementation of such a Conformance tool might be:

```
--ginkgo.focus=\\[Conformance\\]|\\[NodeConformance\\]|\\[sig-windows\\]|\\[sig-apps\\].CronJob --ginkgo.skip=\\[LinuxOnly\\]|\\[k8s.io\\].Pods.*should.cap.back-off.at.MaxContainerBackOff.\\[Slow\\]\\[NodeConformance\\]|\\[k8s.io\\].Pods.*should.have.their.auto-restart.back-off.timer.reset.on.image.update.\\[Slow\\]\\[NodeConformance\\]"
```

> Currently, the defacto standard for windows testing resides at https://github.com/kubernetes-sigs/windows-testing, and we suggest this tag to users, however it should be noted that this does *not* conform to the specification in this document:

Reviewer 2: ```suggestion
Currently, the de facto standard for windows testing resides at https://github.com/kubernetes-sigs/windows-testing, and we suggest this tag to users, however it should be noted that this does *not* conform to the specification in this document:
```

- it triggers some tests which uses non `pause` images for validating functionality.
- it doesn't implement the pod churn or emptyDir sharing tests.
- it doesn't implement the *reboot* tests which are defined in this specification.

> The ability to specify HostProcess, NetworkPolicy, and other tests in this document would follow a similar tagging strategy, and any such tool could just invoke the existing Kuberentes end-to-end test binary with appropriate `focus` and `skip` arguments.

Reviewer 2: ```suggestion
The ability to specify HostProcess, NetworkPolicy, and other tests in this document would follow a similar tagging strategy, and any such tool could just invoke the existing Kubernetes end-to-end test binary with appropriate `focus` and `skip` arguments.
```

#### Option 2 - Sonobuoy implementation for convenience

Given the complexity of the above tagging scheme, we propose a concrete example of a binary program for Windows verification.

```
sonobuoy run --mode=windows-conformance --windows.hostProcess --windows.activeDirectory ...
```

In the above invocation, we verify

- core conformance
- hostProcess

> - activeDirectory 

Reviewer 3: We would need to make sure there was a way to provide AD credentials for any GSMA test

Reviewer 2: We should look at creating automation that sets up AD and any other required domain services so that we can have reproducible builds/tests/conformance.

Reviewer 4: we have scripts that automate this for e2e tests, they wrapped to work with aks-engine but could be extracted out to run on any platform: https://github.com/kubernetes-sigs/windows-testing/tree/master/extensions  

https://testgrid.k8s.io/sig-windows-master-release#aks-engine-windows-containerd-serial-slow-master&include-filter-by-regex=GMSA

functionality for a specific Kubernetes release.  The resulting output would be obtained using

```
sonobuoy status
```

Which would report a binary result of `Pass` or `Fail`.  This tool could thus be used to

> establish a rigorous standard for portable verificatoin of Kubernetes Windows based products across Vendors for both internal and external diagnostic purposes.

Reviewer 2: ```suggestion
establish a rigorous standard for portable verification of Kubernetes Windows based products across Vendors for both internal and external diagnostic purposes.
```

##### Implementation details for sonobuoy

So you'd have something like this:

```
sonobuoy plugin install windows-conformance <url>
sonobuoy run -p windows-conformance
```

The URL can point to any custom URL, but I have a few already made here: https://github.com/vmware-tanzu/sonobuoy-plugins/tree/master/windows-e2e

This is the same as just running from a URL or local file, e.g. `sonobuoy run -p windows-conformance.yaml` or `sonobuoy run -p pluginURL` but I was also showing that we are currently testing out plugin "installation" (caching?) which makes it easier to store, list, use plugins.

> Question: In this scenario (or option 2), Sonobuoy still has to know which tests to invoke. In the end it is still just invoking e2e.test so neither of this really avoid option 1, it does make it easier though.

### Test Plan

This KEP contemplate a full conformance suite for Windows environments already.

### Graduation Criteria

> The plan is to introduct the feature as alpha in the v1.23 time frame having the

Reviewer 2: ```suggestion
The plan is to introduce the feature as alpha in the v1.23 time frame having the
```

initial framework implementation.

#### Alpha -> Beta graduation

- The suite is graduated to beta when the core tests are implemented and running.

#### Beta->GA graduation

> - The suite is graduated to GA when the sub-specifiction tests are implemented and running.

Reviewer 2: ```suggestion
- The suite is graduated to GA when the sub-specification tests are implemented and running.
```

- Ability to delete and recreate pods for StatefulSets which preserve their ability to serve as routed endpoints for services, while also having unchanging IP addresses.
- Ability to access internal pods by internal Kubernetes DNS service endpoints.
- Ability to access external services by Kubernetes DNS services endpoints (for non-airgapped clusters).
- Ability to access internal pods through custom DNS records, injected by the Kubernetes pod Specification.
- Ability to route services from pods from the EndpointSlice API.
- Ability to run IPv4/IPv6 dual-stack networking (on supported OS Versions)

#### Basic Service accounts and local storage

- Ability to rotate service accounts for running pods, with the APIServer access capacity of the pod remaining unchanged over the long term.
- Ability to read and write shared files on a single Kubernetes node between three running Windows containers, simultaneously.

#### Basic Scheduling

- Ability to schedule pods with CPU and Memory limits demonstrably honored over time.
- Ability to demonstrate the pods requesting more CPU and Memory then available are left in the Pending state over time.

#### Basic performance

> 

Reviewer: should we define pods per core....

- Ability to reschedule a deployment Pod, 50 times in a row, with continuous deletions in the background.
- Ability to route traffic to 10 pods behind a common ClusterIP service endpoint.
- Ability to route traffic to 10 pods behind a common NodePort service endpoint.

> We do not include GMSA support in the basic definition of Windows Conformance because it is known to require other cluster components outside of Kubernetes's Control.

Reviewer: @claudiubelu

```

Currently, the de facto standard for windows testing resides at https://github.com/kubernetes-sigs/windows-testing, and we suggest this tag to users, however it should be noted that this does *not* conform to the specification in this document:

- it triggers some tests which uses non `pause` images for validating functionality.
- it doesn't implement the pod churn or emptyDir sharing tests.
- it doesn't implement the *reboot* tests which are defined in this specification.

The ability to specify HostProcess, NetworkPolicy, and other tests in this document would follow a similar tagging strategy, and any such tool could just invoke the existing Kubernetes end-to-end test binary with appropriate `focus` and `skip` arguments.

#### Option 2 - Sonobuoy implementation for convenience

Given the complexity of the above tagging scheme, we propose a concrete example of a binary program for Windows verification.

```

> sonobuoy run --mode=windows-conformance --windows.hostProcess --windows.activeDirectory ...

Reviewer: im not at all committed to this implementation, its a good example though of how we can define windows
operational readiness w/o forcing OS versions, bc the non-standard/non-legacy features can be added on.

Reviewer 4: I like the simplicity of this but we will need somewhere to define what those buckets contain so they can be used across CI and conformance and be tied to a particular k8s release.

Reviewer: i think this is solved since, sonobuoy gives you the full e2e test results/report in junit format right?

#### Story 3

As a Kubernetes developer I want to immediately know whether or not the features I want to add will affect the Windows support matrix, and if so, I want to be able to rapidly convey this in a canonical way to the broader community.

#### Story 4

As an IT administrator, I want to know whether my current version of Kubernetes supports ActiveDirectory

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Specification of Windows Core Conformance (Kubernetes 1.23)

- In order to qualify as a Windows test, the pods that demonstrate functionality for a given attribute of a cluster much be running on a `pause` image, and this image must be scheduled to an operating system that:

    - 1809
    - 1909
    - 2004
    - 2022

> 

Reviewer: @perithompson @rosskirkpat to add clause on 
- patches
- table of known "conformant" OS's

Reviewer 2: I agree with @jsturtevant, it is probably best to track what KBs were present on the test machine rather than trying to maintain a list of KBs that need to be present. This will also make diffs easier if there is an issue

Reviewer: ok, removed this

- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP defines the standard for operationally ready Windows nodes which can support running traditional , Windows workloads.  This a mechanism to solve the end user problem of needing to run Windows workloads on clusters provided by different vendors with potentially different underlying CNI, CRI, or other implementation details.

- Clusters capable of running Windows containers also must be capable of passing Linux Conformance tests, by leveraging available Linux nodes - for example - by leveraging the Linux control plane which is omnipresent in any K8s cluster.

> - Clusters capable of running Windows, with versions > 2019 and forward, conform to the Operations tests defined and standardized in this document.

Reviewer 5: Shouldn't it be `versions >= 2019`?

#### Story 3

As a Kubernetes developer I want to immediately know whether or not the features I want to add will affect the Windows support matrix, and if so, I want to be able to rapidly convey this in a canonical way to the broader community.

#### Story 4

As an IT administrator, I want to know whether my current version of Kubernetes supports ActiveDirectory

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Specification of Windows Core Conformance (Kubernetes 1.23)

- In order to qualify as a Windows test, the pods that demonstrate functionality for a given attribute of a cluster much be running on a `pause` image, and this image must be scheduled to an operating system that:

    - 1809
    - 1909
    - 2004
    - 2022

- All tests in the Core Conformance suite depend on `pause`, and no other images.

> - All tests should use the `Pod.OS` specification field as the *only* required scheduling directive for pods, once this field is made GA.  Up until this time, the way tests schedule Windows pods can use RuntimeClasses, Labels, or taints in any way which "works" to implement the other requirements.

Reviewer 5: s/taints/tolerations. The node is tainted and the pod can tolerate the taint.

Reviewer 4: I think we might need to be more specific in the mean time while the pod.OS field. Not all tests support the various settings and it would good to align them which will make it easier to implement pod.os at some point.

#### Basic concurrent functionality

We intentionally do not define scale targets, as this KEP concerns itself with standardizing minimal functionality required for an operational Windows cluster.

That said, some level of concurrency and repetition is required to verify that a cluster is capable of running functional, multi tenant workloads.

Thus we define the following simplistic "performance" tests.  Since windows service load balancing is not as deeply hardened as that of linux, we also propose that basic scalability requirements which limit node sizes for windows clusters, be put in place, to ensure that "large clusters" are validated in a correct way.

- Ability to reschedule a deployment Pod, 50 times in a row, with continuous deletions in the background.
- Ability to route traffic to 10 pods behind a common ClusterIP service endpoint.
- Ability to route traffic to 10 pods behind a common NodePort service endpoint.

We do not include GMSA support in the basic definition of Windows Conformance because it is known to require other cluster components outside of Kubernetes's Control.

> ### Sub-specifications of Windows Conformance (Kubernetes 1.23)

Reviewer 5: Do we need a sub section for security context to test things like `RunAsUsername`?

Reviewer: added one , thx good point

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP defines an operational readiness standard for Windows nodes which can support running traditional Windows workloads. This is a mechanism to solve the end user problem of needing to run Windows workloads on clusters provided by different vendors with potentially different underlying CNI, CRI, or other implementation and configuration details.

> - Clusters capable of running Windows containers also must be capable of passing Linux Conformance tests, by leveraging available Linux nodes - for example - by leveraging the Linux control plane which is omnipresent in any K8s cluster.

Reviewer: good point, yup, 2 nodes

- Clusters capable of running Windows, with versions > 2019 and forward, conform to the Operations tests defined and standardized in this document.

> In general, the ad-hoc usage of bespoke ginkgo tags is cumbersome and makes adoption of Windows workloads on Kubernetes difficult to certify.  In addition to defining these specific definitional constructs for Windows conformance in this KEP, we also aim to implement these definitions in a declarative, fully automated certification bundle via the sonobuoy (https://sonobuoy.io/) tool, as a quick and easy way for end-users to verify windows clusters.

Reviewer: ah yeah split up and updated now

- Some tests (including Networking, NetworkPolicy, and Storage related tests) behave differently on Windows because of differences in the feature sets and testing idioms that are supportable on Windows Server. We aim to distinguish the core requirements of such tests in terms of conformant Windows clusters regardless of existing implementation gaps which may exist between how Linux and Windows tests are implemented.
- Some critical Windows functionality, such as the ability to run activeDirectory as an identity provider for a pod, needs to be well-defined in a Kubernetes context for individuals planning on integrating production Windows workloads into their Kubernetes support model.
- The introduction of "Privileged containers" (formally referred to as `HostProcess` pods) further differentiates Linux from Windows in subtle ways which will allow Windows users to implement many of the common idioms in Kubernetes using a comparable runtime paradigm with slight differences.

> - The Kubernetes ServiceProxy implementation is not at parity with that of Linux and is potentially going to lag behind for the foreseeable future. This will need to be explicitly defined to allow future Kubernetes service API implementation parity attempts on Windows and Linux to be approached in a systematic manner in terms of core requirements and supportability. Certain features may require special consideration for Windows clusters which may have unintended interactions with the underlying CNI provider. Some examples of these additional layers of complexity are present in current features such as DSR (Direct Server Return) and EndpointSlices.

Reviewer: added DSR and kube proxy notes, also added the stuff about runas username

#### Story 4

As an IT administrator, I want to know whether my current version of Kubernetes supports ActiveDirectory or other Windows features.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Specification of Windows Operational Readiness (Kubernetes 1.23)

- In order to qualify as a Windows test, the pods that demonstrate functionality for a given attribute of a cluster must be running on a `pause` image. 

Additionally, this image must be scheduled to an operating system that meets the following requirements:

- Windows Build Versions:
  - Windows Server 2019 (LTSC)
  - Windows Server 2022 (LTSC)

- Installed KBs:
  <!-- - Needs further discussion/investigation -->

> - Windows Hosts and containers must be running a patch level that is no older than 3 months

Reviewer: moved out now...

> - All tests in the Core Windows Operational Readiness suite depend on `pause`, and no other images.

Reviewer 4: Not sure I understand Why only the pause image?

> - All tests in the Core Windows Operational Readiness suite depend on `pause`, and no other images.

Reviewer: oh wait, sorry i meant agnhost !

> - All tests in the Core Windows Operational Readiness suite depend on `pause`, and no other images.

Reviewer: fixed

- All tests should use the `Pod.OS` specification field as the *only* required scheduling directive for pods, once this field is made GA.  Up until this time, the way tests schedule Windows pods can use RuntimeClasses, Labels, or taints in any way which "works" to implement the other requirements.

> #### Basic Networking

Reviewer 4: These are lists of functionally, has there been any mapping to the current Conformance tests that run on windows and Windows specific tests to find gaps in functionality that isn't tested for windows?

Reviewer: yeah, these are reverse engineered from the tests we already have,  ive manually confirmed we have tags for these.

- Ability to mount a hostPath storage volume from a Kubelet into a pod and write to it. 
- Ability to mount a hostPath storage volume from a Kubelet into a pod and read from it. 
- Ability to schedule multiple containers, with distinct IP addresses, on the same node.
- Ability to reboot all of the Windows nodes in a cluster while running Windows pods, in such a way that recovery of these deleted pods occurs in the scope of a few minutes.
- Ability to delete and recreate services in such a way that load balancing rules for pods are recovered by whatever chosen service proxy is being utilized.
- Ability to delete and recreate pods for StatefulSets which preserve their ability to serve as routed endpoints for services, while also having unchanging IP addresses.
- Ability to access internal pods by internal Kubernetes DNS service endpoints.
- Ability to access external services by Kubernetes DNS services endpoints (for non-airgapped clusters).
- Ability to access internal pods through custom DNS records, injected by the Kubernetes pod Specification.
- Ability to route services from pods from the EndpointSlice API.

> - Ability to run IPv4/IPv6 dual-stack networking (on supported OS Versions)

Reviewer 4: dual-stack is not support in 2019 and there are no plans to backport it as of now.

Reviewer: fixed

- Ability to support IPv6 and IPv4 services for accessing internal Windows pods from Linux pods over ClusterIP Endpoints.
- Ability to support IPv6 and IPv4 services for accessing internal Windows pods from Linux pods over NodePort Endpoints.

### Implementation

This section details the implementation of a "tool" which verifies these clusters must allow for the following, differentiated set of verifications, specified above.

- core conformance
- hostProcess
- activeDirectory
- networkPolicy
- serviceProxy

#### Option 1 - Test tags

One way to implement such a tool is to, specifically, target all available "ginkgo" tests in Kubernetes core itself, and run these one at a time (or concurrently).  Not we *do not* assert that these tests *must* run in parallel in order to comprise a conformance clusters (although this is often done in CI, its generally not a requirement of Kubernetes Conformance for linux clusters, either, since cluster capacity is not related to the functionality of a cluster). As an example, one such implementation of such a Conformance tool might be:

```

> --ginkgo.focus=\\[Conformance\\]|\\[NodeConformance\\]|\\[sig-windows\\]|\\[sig-apps\\].CronJob --ginkgo.skip=\\[LinuxOnly\\]|\\[k8s.io\\].Pods.*should.cap.back-off.at.MaxContainerBackOff.\\[Slow\\]\\[NodeConformance\\]|\\[k8s.io\\].Pods.*should.have.their.auto-restart.back-off.timer.reset.on.image.update.\\[Slow\\]\\[NodeConformance\\]"

Reviewer 4: We also run a who suite of slow/serial tests, some of them windows only. We should ensure we included those as well.

Reviewer: fair point, unskipped `slow`

#### Basic concurrent functionality

We intentionally do not define scale targets, as this KEP concerns itself with standardizing minimal functionality required for an operational Windows cluster.

That said, some level of concurrency and repetition is required to verify that a cluster can be functional, multi-tenant workloads.

Thus, we define the following simplistic "performance" tests.  Since Windows service load balancing is not as deeply hardened as that of Linux, we also propose that basic scalability requirements which limit node sizes for Windows clusters, be put in place, to ensure that "large clusters" are validated in a correct way.

- Ability to reschedule a deployment Pod, 50 times in a row, with continuous deletions in the background.
- Ability to route traffic to 10 pods behind a common ClusterIP service endpoint.
- Ability to route traffic to 10 pods behind a common NodePort service endpoint.

We do not include GMSA support in the basic definition of Windows OR because it is known to require other cluster components outside of Kubernetes's Control.

> ### Sub-specifications of Windows Operational Readiness (Kubernetes 1.24)

Reviewer 6: Should we add some checks that ensure that either Defender is disabled or if it is there are some well-known exclusions set up?
We know for sure that not including a windows defender exclusion for processes like containerd can lead to very poor performance.

I'm not sure where best to add this sort of check since it isn't really testing any K8s functionality (but i think is still important)

Reviewer 2: ```powershell
# disable certain features of microsoft defender to increase the speed of running/setting Kubernetes on Windows
Set-MpPreference -DisableRealtimeMonitoring $true -DisableScriptScanning $true -DisableArchiveScanning $true 
Set-MpPreference -ExclusionPath <BIN_PATH>

# verify our defender preferences were set
Get-MpPreference
Get-MpComputerStatus
```

These are my standard go-to commands for dealing with Defender. The exclusion path needs to be set based on what you use for your binary path(s). We can also add exclusions for specific processes, with an example below. Assuming we can use powershell to add the checks, this should be very easy to adapt.

```powershell
$processNames = @( 'containerd.exe','kubelet.exe','kube-proxy.exe','calico.exe', 'calico-node.exe', 'calico-ipam.exe' )
    ForEach ($p in $processNames) {
        Get-CimInstance -ClassName Win32_Process -Filter  "name = '$p'" | Select-Object -First 1 | ForEach-Object { 
                Set-MpPreference -ExclusionProcess $($_.Path)
        }
}
```

Author: I added an item as `Windows Worker Configuration`, the command can be used as part of the test, not sure if we should add in the KEP since there isn't formal examples for the other items.. @rosskirkpat

- All tests should use the `Pod.OS` specification field as the *only* required scheduling directive for pods, once this field is made GA.  Up until this time, the way tests schedule Windows pods can use RuntimeClasses, Labels, or taints in any way which "works" to implement the other requirements.
  - until this field is fully available in all tests, the use of  `tolerations` and `labels` should be used for all tests to solidify the scheduling aspects of windows containers. 
 
#### Basic Networking

- Ability to access Linux container IPs *by service IP* from Windows containers
- Ability to access Windows container IPs *by service IP* from Linux containers
- Ability to access Linux container IPs *by NodePort IP* from Windows containers
- Ability to access Windows container IPs *by NodePort IP* from Linux containers
- Ability to access Linux container IPs *by pod IP* from Windows containers
- Ability to access Windows container IPs *by pod IP* from Linux containers
- Ability to schedule multiple containers, with distinct IP addresses, on the same node.

> - Ability to delete and recreate services in such a way that load balancing rules for pods are recovered by whatever chosen service proxy is being utilized.

Reviewer 2: Is there a reason we are not specifying kube-proxy as the service proxy here?

Reviewer 2: Additionally, is the vagueness here surrounding the service proxy choice intentional? I believe that we should initially standardize on kube-proxy for Windows as it is the only viable **and** currently available service proxy for Windows that I am aware of. Leaving it as-is could result in anyone falling into our user stories above to interpret this as _any_ service proxy drop in can pass our Windows Operational Readiness conformance tests.

Reviewer 2: @jayunit100 @jsturtevant @knabben any thoughts?

Reviewer 4: I think the most used proxy is winkernel (used by azurecni, calico, flannel, and win-brdige) but I believe Antrea uses the userspace proxy.  

If someone were to develop a proxy that met all these requirements and passed the test is there a reason to block them from passing these readiness tests? Sounds similiar to how you can use any cni as long as it conformance to a base set of rules.

Reviewer 2: Thanks for the clarification! Apologies about that, was doing too many things at once and had a short brain lapse when reading the service proxy portions.

Author: We have kpng as well, so agree on having an open statement about the proxy here

- The introduction of "Privileged containers" (formally referred to as `HostProcess` pods) further differentiates Linux from Windows in subtle ways which will allow Windows users to implement many of the common idioms in Kubernetes using a comparable runtime paradigm with slight differences.
- The Kubernetes "kube-proxy" implementation is not at parity with that of Linux and is potentially going to lag behind for the foreseeable future. For example, some features (such as EndpointSlices) aren't implemented in the Windows userspace proxy.  Yet other features such as DSR and terminating endpoints have specific Windows behaviors which aren't identical to that of the Linux service proxy https://github.com/kubernetes/kubernetes/issues/96514.
- The way users run pods on Windows, as a specific user, varies from the implementation on Linux, as illustrated here: https://github.com/openshift/windows-machine-config-operator/pull/638. 

> We propose the following definitions as the Windows conformance standard for Kubernetes clusters that are released after Kubernetes 1.24 and onwards. This specification is to be updated as new features are added or removed from the Kubernetes API itself.

Reviewer: it could be , but thats not addressed in this KEP (feel free to add it as a sentence if you think it should be?)

> We propose the following definitions as the Windows conformance standard for Kubernetes clusters that are released after Kubernetes 1.24 and onwards. This specification is to be updated as new features are added or removed from the Kubernetes API itself.

Reviewer 4: ok, this is what I thought "We propose the following definitions as the Windows conformance standard for Kubernetes clusters" makes it sound like we are providing conformance.  maybe want to change to 

```suggestion
We propose the following definitions as the Windows Operational Readiness for Kubernetes clusters that are released after Kubernetes 1.24 and onwards. This specification is to be updated as new features are added or removed from the Kubernetes API itself.
```
```suggestion
We propose the following definitions as the Windows conformance standard for Kubernetes clusters that are released after Kubernetes 1.24 and onwards. This specification is to be updated as new features are added or removed from the Kubernetes API itself.
```

### User Stories (Optional)

#### Story 1

As a Windows Application developer, I want to verify that the features I rely on in Kubernetes are supported on my specific Windows Kubernetes cluster.

#### Story 2

As a Kubernetes vendor, I want to evaluate the completeness of my Kubernetes support matrix in the context of Windows supported features in the Kubernetes API.

#### Story 3

As a Kubernetes developer, I want to immediately discern whether the features I want to add will affect the Windows support matrix. If there are any confirmed or potential impact(s) I want to be able to rapidly convey them in a canonical way to the broader community.

#### Story 4

As an IT administrator, I want to know whether my current version of Kubernetes supports ActiveDirectory or other Windows features.

### Risks and Mitigations

## Design Details

### Specification of Windows Operational Readiness (Kubernetes 1.24)

> - In order to qualify as a Windows test, the pods that demonstrate functionality for a given attribute of a cluster must be running on a `pause` image. 

Reviewer 4: I am not sure I understand this requirement? Technically the pause image is a implementation detail and should not be a requirement

Reviewer: wait, i thought we deleted this, it was a typo, supposed to be `agnhost` :)

Reviewer: ill fix it

Reviewer: fieddd

Reviewer 4: > IMO image drift complicates e2e testing

agree, we should try to keep it to the main images anghost/pause/busybox

Additionally, this image must be scheduled to an operating system that meets the following requirements:

- Windows Build Versions:
  - Windows Server 2019 (LTSC)
  - Windows Server 2022 (LTSC)

- Installed KBs:

>   - Kubernetes 1.21 and forward

Reviewer 4: This should be something like a windows KB?

Reviewer: i think we decided against having the KBs in here.  deleting this.

> - All tests in the Core Windows Operational Readiness suite depend on `pause` as well as `agnhost`, and no other images.  This is because there are no requirements (thus far) in this specification which cannot be tested via the `pause` image, and thus it simplifies the types of pods we test on.  In the future we may changes this requirement if new workloads are important for operational readiness specification (i.e. a GPU intensive app or a CNI for host process verification).

Reviewer 4: we also rely on busybox, etcd, and the private pull images (some of which have sub-dependencies) all can be built out of the https://github.com/kubernetes/kubernetes/tree/master/test/images folder

Reviewer: updating shortly to have more nuance and accomodate many different types of images, but on a case by case basis, contextualizing busybox/etcd appropriately as transitive, non workload images.

> - All tests should use the `Pod.OS` specification field as the *only* required scheduling directive for pods, once this field is made GA.  Up until this time, the way tests schedule Windows pods can use RuntimeClasses, Labels, or taints in any way which "works" to implement the other requirements.

Reviewer 4: pod.OS being used as scheduling is still up in the air afaik cc @ravisantoshgudimetla

Reviewer: fixed, pushing shortly

- Ability to protect IPv4 pods from accessing other pods when TCP NetworkPolicies are present that block specific pod connectivity.
- Ability to protect IPv4 pods from accessing other pods when TCP NetworkPolicies are present that block specific namespace connectability.

#### Windows Advanced Networking and Service Proxying

These apply to Windows Server 2019 and up (we expect windows 2004 to be out of support once this KEP merges).

- Ability to support IPv6 and IPv4 services for accessing internal Linux pods from Windows pods over ClusterIP Endpoints.
- Ability to support IPv6 and IPv4 services for accessing internal Linux pods from Windows pods over NodePort Endpoints.
- Ability to support IPv6 and IPv4 services for accessing internal Windows pods from Linux pods over ClusterIP Endpoints.
- Ability to support IPv6 and IPv4 services for accessing internal Windows pods from Linux pods over NodePort Endpoints.
- Ability to run IPv4/IPv6 dual-stack networking (on supported OS Versions).

Note that, ipv4 and ipv6 are both known to be highly CNI (overlay vs non-overlay) and OS (2019 vs 2022) dependent.

> #### Windows Worker Configuration

Reviewer 4: Is there more than this? things like the variables features enabled, ssh, versions of the kbs installed, windows updates disabled/enabled

Reviewer 4: how are these going to be verified? We don't have any tests that do this today?

Reviewer: adding asection about how these arent actually automatically verified. 
I didnt add KBs but i did add RDP/SSH details.
comimting shortly

Workers node settings that must be compliant.

> - Microsoft Defender must be disabled in the Windows workers.

Reviewer 4: why are recommending disabling Defender?

Reviewer 6: Ooops, I missed that.
We should ensure that there are defender exclusions set for things like containerd but we should not require that it be disabled. (at least in my opinion)

Author: my mistake, enforcing the exclusion of certain processes, contained for now.
 @jsturtevant @marosset PTAL

### Implementation

This section details the implementation of a "tool" which verifies these clusters must allow for the following, differentiated set of verifications, specified above.

- Core Conformance
- HostProcess
- ActiveDirectory
- NetworkPolicy
- ServiceProxy

As an example, if a Windows verification run for 1.22 was occurring, the highest semver close to 1.22, could be used
for running NetworkPolicy verification.  In this case, 1.21.  This would allow for backward compatibility of this file
so that if a ginkgo tag changed, a new "test" with a new "version" and the SAME operationalReadinessDescription could be added.

As part of this KEP, we will finish the ongoing ginkgo tag curation at https://docs.google.com/spreadsheets/d/1Pz7-AUZ9uxMBwFx7ZC2U6dBfYpMUdOUi1UivBRVDw7I/edit#gid=0 - to live in the a `sig-windows` repository as an reference summarization and a convinient grouping of the tests.

> This KEP implements a layer of abstraction mainly in a YAML/JSON, which defines each item described below containing

Reviewer 4: how is this yaml going to be consumed by sonobouy or e2e.test?  How is maintained? this looks fairly manual to list every test

Reviewer: Yup, its definetly manual.  
No plans for long term maintainance.  Its a one time task to define the operational readiness tags for 1.23/24 really.  After that, like anything else, the update lifecycle process for it will depend on how well it serves the hard requirements of the KEP's actual deliverables.

Reviewer 4: The prototype answered many of my questions here.  We do need to think about how we version these json files per release but we can do that for moving to beta

#### Option 2 - Sonobuoy implementation for convenience

Given the complexity of the above tagging scheme, we propose a concrete example of a binary program for Windows verification. 
This was the option chosen for this KEP implementation.

```
sonobuoy run --mode=windows-conformance --windows.hostProcess --windows.activeDirectory ...
# which prints out the exact tests so that the actual test coverage is clear
```

In the above invocation, we verify

- Core Conformance
- HostProcess
- ActiveDirectory 

functionality for a specific Kubernetes release. The resulting output would be obtained using

```
sonobuoy status
```

Which would report a binary result of `Pass` or `Fail`. This tool could thus be used to establish a rigorous standard for portable verification of Kubernetes Windows based products across Vendors for both internal and external diagnostic purposes.

##### Implementation details for Sonobuoy

Below is an example of running the Windows Operational Readiness conformance tests with Sonobuoy

```
sonobuoy plugin install windows-operational-readiness <url>
sonobuoy run -p windows-operational-readiness
```

> The URL can point to any custom URL. We have published a few examples of Windows e2e tests here: https://github.com/vmware-tanzu/sonobuoy-plugins/tree/master/windows-e2e

Reviewer 4: where will this canonical list live?  Since the test definitions are in these configurations they could run any set of tests configured and call it good.

Reviewer: updated to point to the GA section as the final place we put the endpoint for all these artifacts.

### Test Plan

This KEP defines the test plan for such a cluster, but in general, we will update all of our CI jobs to implement it.

### Graduation Criteria

The plan is to introduce the feature as alpha in the v1.24 time frame having the
initial framework implementation.

At this time, we'll expect that users can either:

1) Leverage a simple process to run specific tests validating a feature, such as...
```
	curl coreconformance.1.23.yaml
	skiptags=`cat coreconformance1.23.yaml` | yq .basicServices.reboot_all_nodes | grep skipTags
	skiptags=`cat coreconformance1.23.yaml` | yq .basicServices.reboot_all_nodes | grep focusTags
	e2e.test -skip="$skiptags" -focus="$focustags"
```

2) Run `sonobuoy` command line tooling, which consumes the above raw daya, to do the same via
command line flags:

```
	
```

#### Alpha -> Beta graduation

- The suite is graduated to beta when the core tests are implemented and running.

#### Beta -> GA graduation

- The suite is graduated to GA when the sub-specification tests are implemented and running.

### Upgrade / Downgrade Strategy

### Version Skew Strategy

N/A

> ## Production Readiness Review Questionnaire

Reviewer 7: This KEP appears to be about building a test suite as opposed to a feature that is enabled and always present inside of a running cluster.  If I've understood that correctly, then the PRR answers aren't required.

To handle this case, explain why PRR does not apply in this section and remove the rest of question.  Here's an example https://github.com/kubernetes/enhancements/pull/3137/files#diff-0ad9f743dbfdf3932906414acf0a79e074af723ad2ffb8e0dab3429f0b664beeR249-R252

Then I'll approve the PRR

Author: @deads2k Squashing commits and adding the PRR sections clarification

Reviewer: sounds like were all set now

### Test Plan

This KEP defines the test plan for such a cluster, but in general, we will update all of our CI jobs to implement it.

### Graduation Criteria

The plan is to introduce the feature as alpha in the v1.24 time frame having the
initial framework implementation.

At this time, we'll expect that users can either:

1) Leverage a simple process to run specific tests validating a feature, such as...
```
	curl coreconformance.1.23.yaml
	skiptags=`cat coreconformance1.23.yaml` | yq .basicServices.reboot_all_nodes | grep skipTags
	skiptags=`cat coreconformance1.23.yaml` | yq .basicServices.reboot_all_nodes | grep focusTags
	e2e.test -skip="$skiptags" -focus="$focustags"
```

2) Run `sonobuoy` command line tooling, which consumes the above raw daya, to do the same via
command line flags:

```
	
```

#### Alpha -> Beta graduation

- The suite is graduated to beta when the core tests are implemented and running.

#### Beta -> GA graduation

- The suite is graduated to GA when the sub-specification tests are implemented and running.

### Upgrade / Downgrade Strategy

### Version Skew Strategy

N/A

## Production Readiness Review Questionnaire

> Not applicable because this is a policy KEP.

Reviewer 6: ```suggestion
Not applicable because this KEP is focused on building Windows focused test suite.
```
